### PR TITLE
RCHAIN-3888: Secure access to protocol-private PoS methods

### PR DESCRIPTION
**Pull request by tymm**
## Overview
See: https://gist.github.com/tymm/f1b89ef2d3308ccaec858b16937e6807 and https://rchain.atlassian.net/browse/RCHAIN-3888

I would appreciate if @marcin-rzeznicki and @JosDenmark could take a closer look on what I did here with replaying system deploys from `BlockMessage`.
Also please have a look on how I use the `invalidBlockHash` as a seed for the rand value for `SlashDeploy`s.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by tymm** [comment]
This PR is blocked until I can replace the following code with a system deploy.
https://github.com/rchain/rchain/blob/179571510a48fe6d5338f8d98a0e232ab0917192/casper/src/main/scala/coop/rchain/casper/BlockCreator.scala#L65-L77

**Comment by marcin-rzeznicki** [comment]
Hey @tymm I'll finish slashing system deploy today to unblock you, thx

**Comment by marcin-rzeznicki** [comment]
> This PR is blocked until I can replace the following code with a system deploy.
> 
> https://github.com/rchain/rchain/blob/179571510a48fe6d5338f8d98a0e232ab0917192/casper/src/main/scala/coop/rchain/casper/BlockCreator.scala#L65-L77

This is fixed in 9e562ab (once it's merged)

**Review by marcin-rzeznicki — APPROVED** [review]
:+1:

**Comment by tymm** [comment]
bors try

**Comment by tymm** [comment]
bors try

**Comment by woky** [comment]
bors try

**Comment by tymm** [comment]
bors try

**Comment by JosephDenman** [comment]
@tymm Would you explain why this is something that a system `DeployerID` can't address? Or send me a link to our previous discussion with Artur?

**Comment by tymm** [comment]
@JosDenmark the discussion can be found in this gist: https://gist.github.com/tymm/f1b89ef2d3308ccaec858b16937e6807

File: casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala]: If there are no expected system deploys in the genesis block, we should verify that the block contains none.

File: casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala]: Why add a parameter instead of using the pre-existing `replaySystemDeploy` function?

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala]: Isn't the `override` redundant since `rand` is passed to `SystemDeploy`?

Comment (JosephDenman) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala]: And the `val` modifier? I don't see it accessed anywhere outside of `SystemDeploy`.

File: casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala
Comment (JosephDenman) [reply casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala]: Like you did for the `computeDeploysCheckpoint` function.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala]: Appending to a list is an anti-pattern. You should use `Vector`.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala]: The error message should be extracted and added to the `Exception` message.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala:92]: This is an attack vector, since we're blindly running system deploys that a potentially malicious validator has sent us. We need to recalculate the slashing deploys on the replay side and verify that they match the system deploys sent in the block message. Then we replay them.

File: models/src/main/scala/coop/rchain/casper/protocol/CasperMessage.scala
Comment (JosephDenman) [inline models/src/main/scala/coop/rchain/casper/protocol/CasperMessage.scala]: Why not just have `SlashSystemDeployData` extend `SystemDeployData`? Then `playSucceeded` is subtype polymorphic over the `SystemDeployData`. More extensible.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala]: Again, I'm unsure why you would want to use a different RNG than the one passed to `SystemDeploy`.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala:444]: I don't think either of the `replayDeploy` or `replaySystemDeploy` parameters are necessary, since we use the `isGenesis` flag in `replayComputeState`. The first existed before this PR, so whether you want to remove it is up to you. `replaySystemDeploy` should probably be removed though.

**Review by JosephDenman — DISMISSED** [review]
Good work overall. I'm only up to the fourth commit (which I think should've been its own PR), but I want you to respond to some of my comments before I continue.

File: casper/src/main/resources/PoS.rhox
Comment (tgrospic) [inline casper/src/main/resources/PoS.rhox]: Is this comment still valid?

File: casper/src/main/resources/PoS.rhox
Comment (tgrospic) [inline casper/src/main/resources/PoS.rhox]: Shouldn't this return the same Either type e.g. `return!((false, "Invalid auth token"))`? In other methods also.

File: casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala]: Good call!

File: casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala]: Because the pre-existing `replaySystemDeploy(startHash: StateHash)(...)` function does not fit well into `RuntimeManager.replayComputeState` which is unaware of the the start state hashes of each deploy or system deploy.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala]: The `override val` was necessary since `SystemDeploy` has a val called "rand".
But now that you said that, I realized that I don't need `override val` when I use an identifier different from "rand". Thanks.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala:92]: The malicious validator would slash him or herself in that case. This is what ```getInvalidBlocks(`rho:casper:invalidBlocks`)``` in `PoS!slash` ensures.
I didn't know this myself before consulting Kent about it.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala]: This was related to what I described here: https://github.com/rchain/rchain/pull/2781#discussion_r365625061

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala:444]: Are you saying the parameters can be removed because `replayComputeState` is only used for genesis?

File: casper/src/main/resources/PoS.rhox
Comment (tymm) [reply casper/src/main/resources/PoS.rhox]: Good catch!

File: casper/src/main/resources/PoS.rhox
Comment (tymm) [reply casper/src/main/resources/PoS.rhox]: Done.

File: casper/src/main/resources/PoS.rhox
Comment (tgrospic) [inline casper/src/main/resources/PoS.rhox:344]: It seems that `slash` is returning just Boolean (line 330).

File: casper/src/main/resources/PoS.rhox
Comment (tgrospic) [inline casper/src/main/resources/PoS.rhox:534]: `closeBlock` is returning `Nil` for success. Maybe it makes sense to change `slash` and `closeBlock` to return Either as other methods.

**Comment by tymm** [comment]
bors try

**Comment by tymm** [comment]
bors try

**Review by tgrospic — APPROVED** [review]

**Comment by tymm** [comment]
> bors r+

@9rb please don't just try to merge PRs without explicit approval from the creator.

**Comment by tymm** [comment]
I think this PR would really need further input from Joe but Rao wants this to be merged.
bors r+

**Pull request merged**
